README.md
# Mashpack

Mashpack is a JSON-object serialization and compression specification
inspired by [MessagePack](https://msgpack.org)
but is tweaked for optimizing for common JSON objects and for data
structures and layouts used commonly in data sciences.

There are a few notable changes between Mashpack and MessagePack along with
explanations as to why these changes were made:

- Mashpack changes the order of the prefixing for data types from what Messagepack
  would describe as it's 'fixed' data types to be spread more evenly across
  common data types that are used in almost even JSON object.

  These prefixed data types are the ones that can be described with only a single
  byte of header information and are constrained to a range of values to maintain
  this property. The more data that can fit into these types the less often
  the packer will have to add additional header bytes.

  | Spec        | Prefixes                                                                   |
  |-------------|----------------------------------------------------------------------------|
  | Mashpack    | `MAPP(2), STRP(2) MARRAYP(3) INTP(3)`                                      |
  | Messagepack | `positive fixint(1), fixstr(3), negative fixint(3), fixmap(4) fixarray(4)` |

  | Data Type             | Range in Mashpack  | Range in Messagepack |
  |-----------------------|--------------------|----------------------|
  | `MAPP vs fixmap`      | 0 to 63 key-values | 0 to 15 key-values   |
  | `STRP vs fixstr`      | 0 to 63 bytes      | 0 to 31 characters   |
  | `MARRAYP vs fixarray` | 0 to 31 elements   | 0 to 15 elements     |
  | `NINTP vs nfixint`    | -1 to -32          | -1 to -32            |
  | `INTP vs pfixint`     | 0 to 31            | 0 to 127             |

- Mashpack defaults to 'typed' arrays which all have the name `ARRAY*`. These
  function, pack, and unpack exactly the same as mixed arrays except they
  take advantage of having the same type of object within them in order to
  compress tighter together only requiring the type information of all
  objects within the array at the beginning of the array. In most cases arrays
  will all be of similar type and so this situation happens quite frequently
  within common JSON objects.
  
  Typed arrays are allowed to do type conversion of `*P` objects into their `*8`
  counterparts. For example:
  
  ```
  [1, 255, 255, 255...] would be converted to MARRAY8[INTP[1], INT8[255], INT8[255], INT8[255]...]
  but could be converted to ARRAY8[INT8[1], INT8[255], INT8[255], INT8[255]...] by changing the sole INTP to INT8.
  ```

- To use an array with mixed element types the `MARRAY*` (mixed array) data type
  is used. This carries a compression penalty that puts array size in-line with
  Messagepack's arrays.

- Mashpack sheds a lot of the `EXT*` data types that are used in Messagepack in favor
  of just three: `EXT8`, `EXT16`, and `EXT32`. Mashpack reserves all `EXT` codes that have a `1`
  in the most significant bit of their extension code.
  
The majority of the code in this repository is sourced from or directly inspired from
[`msgpack/msgpack-python`](https://github.com/msgpack/msgpack-python). Files are listed
with the copyright and license information of both the original Messagepack code and
the changes. Both projects are licensed under Apache-2.0. See LICENSE.

## Specification

All multi-byte integers are packed with network byte-order.

### Data Types Overview

| Data Type | Prefix     | First Byte  |
|-----------|------------|-------------|
| MAPP      | `00xxxxxx` | `0x00-0x3F` |
| STRP      | `01xxxxxx` | `0x40-0x7F` |
| MARRAYP   | `100xxxxx` | `0x80-0x9F` |
| INTP      | `101xxxxx` | `0xA0-0xBF` |
| NINTP     | `111xxxxx` | `0xE0-0xFF` |
| FALSE     | `11000000` | `0xC0`      |
| TRUE      | `11000001` | `0xC1`      |
| MAP8      | `11000010` | `0xC2`      |
| MAP16     | `11000011` | `0xC3`      |
| MAP32     | `11000100` | `0xC4`      |
| STR8      | `11000101` | `0xC5`      |
| STR16     | `11000110` | `0xC6`      |
| STR32     | `11000111` | `0xC7`      |
| ARRAY8    | `11001000` | `0xC8`      |
| ARRAY16   | `11001001` | `0xC9`      |
| ARRAY32   | `11001010` | `0xCA`      |
| MARRAY8   | `11001011` | `0xCB`      |
| MARRAY16  | `11001100` | `0xCC`      |
| MARRAY32  | `11001101` | `0xCD`      |
| BIN8      | `11001110` | `0xCE`      |
| BIN16     | `11001111` | `0xCF`      |
| BIN32     | `11010000` | `0xD0`      |
| INT8      | `11010001` | `0xD1`      |
| INT16     | `11010010` | `0xD2`      |
| INT32     | `11010011` | `0xD3`      |
| INT64     | `11010100` | `0xD4`      |
| UINT8     | `11010101` | `0xD5`      |
| UINT16    | `11010110` | `0xD6`      |
| UINT32    | `11010111` | `0xD7`      |
| UINT64    | `11011000` | `0xD8`      |
| FLOAT32   | `11011001` | `0xD9`      |
| FLOAT64   | `11011010` | `0xDA`      |
| EXT8      | `11011011` | `0xDB`      |
| EXT16     | `11011100` | `0xDC`      |
| EXT32     | `11011101` | `0xDD`      |
| RESERVED  | `11011110` | `0xDE`      |
| NULL      | `11011111` | `0xDF`      |

### Map Family (`MAPP`, `MAP8`, `MAP16`, `MAP32`)

```
MAPP stores a mapping of up to 63 key-value pairs
+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|00XXXXXX| N*2 Mashpack-encoded objects |
+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXX is a 6-bit unsigned integer

MAP8 stores a mapping of up to 255 key-value pairs
+--------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xC2  |XXXXXXXX| N*2 Mashpack-encoded objects |
+--------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXXXX is a 8-bit unsigned integer

MAP16 stores a mapping of up to 65,535 key-value pairs
+--------+----------------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xC3  |XXXXXXXXXXXXXXXX| N*2 Mashpack-encoded objects |
+--------+----------------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXXXX_XXXXXXXX is a 16-bit unsigned integer

MAP32 stores a mapping of up to 4,294,967,295 key-value pairs
+--------+--------------------------------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xC3  |XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX| N*2 Mashpack-encoded objects |
+--------+--------------------------------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is a 32-bit unsigned integer
```

### String Family (`STRP`, `STR8`, `STR16`, `STR32`)

```
STRP stores a UTF-8 encoded string up to 63 bytes
+--------+======================+
|01XXXXXX| UTF-8 encoded string |
+--------+======================+
where XXXXXX is a 6-bit unsigned integer

STR8 stores a UTF-8 encoded string up to 255 bytes
+--------+--------+======================+
|  0xC5  |XXXXXXXX| UTF-8 encoded string |
+--------+--------+======================+
where XXXXXX is a 8-bit unsigned integer

STR16 stores a UTF-8 encoded string up to 65,535 bytes
+--------+----------------+======================+
|  0xC6  |XXXXXXXXXXXXXXXX| UTF-8 encoded string |
+--------+----------------+======================+
where XXXXXX_XXXXXXXX is a 16-bit unsigned integer

STR32 stores a UTF-8 encoded string up to 4,294,967,295 bytes
+--------+--------------------------------+======================+
|  0xC7  |XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX| UTF-8 encoded string |
+--------+--------------------------------+======================+
where XXXXXX_XXXXXXXX_XXXXXX_XXXXXXXX is a 32-bit unsigned integer
```

### Mixed Array Family (`MARRAYP`, `MARRAY8`, `MARRAY16`, `MARRAY32`)

```
MARRAYP stores a mapping of up to 63 elements
+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|100XXXXX| N Mashpack-encoded objects |
+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXX is a 5-bit unsigned integer

MARRAY8 stores a mapping of up to 255 elements
+--------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xCB  |XXXXXXXX| N Mashpack-encoded objects |
+--------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXXXX is a 8-bit unsigned integer

MARRAY16 stores a mapping of up to 65,535 elements
+--------+----------------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xCC  |XXXXXXXXXXXXXXXX| N Mashpack-encoded objects |
+--------+----------------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXXXX_XXXXXXXX is a 16-bit unsigned integer

MARRAY32 stores a mapping of up to 4,294,967,295 elements
+--------+--------------------------------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xCD  |XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX| N Mashpack-encoded objects |
+--------+--------------------------------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is a 32-bit unsigned integer
```

### Typed Array Family (`ARRAY8`, `ARRAY16`, `ARRAY32`)

```
ARRAY8 stores a mapping of up to 255 elements
+--------+--------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xC8  |XXXXXXXX|YYYYYYYY| N Mashpack-encoded objects |
+--------+--------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXXXX is a 8-bit unsigned integer
and YYYYYYYY is the header byte of all encoded elements.

ARRAY16 stores a mapping of up to 65,535 elements
+--------+----------------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xC9  |XXXXXXXXXXXXXXXX|YYYYYYYY| N Mashpack-encoded objects |
+--------+----------------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXXXX_XXXXXXXX is a 16-bit unsigned integer
and YYYYYYYY is the header byte of all encoded elements.

ARRAY32 stores a mapping of up to 4,294,967,295 elements
+--------+--------------------------------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xCA  |XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX|YYYYYYYY| N Mashpack-encoded objects |
+--------+--------------------------------+--------+~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is a 32-bit unsigned integer
and YYYYYYYY is the header byte of all encoded elements.
```

### Binary Family (`BIN8`, `BIN16`, `BIN32`)

```
BIN8 stores a byte string up to 255 bytes
+--------+--------+=======+
|  0xCE  |XXXXXXXX| bytes |
+--------+--------+=======+
where XXXXXX is a 8-bit unsigned integer

BIN16 stores a byte string up to 65,535 bytes
+--------+----------------+=======+
|  0xCF  |XXXXXXXXXXXXXXXX| bytes |
+--------+----------------+=======+
where XXXXXX_XXXXXXXX is a 16-bit unsigned integer

BIN32 stores a byte string up to 4,294,967,295 bytes
+--------+--------------------------------+=======+
|  0xD0  |XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX| bytes |
+--------+--------------------------------+=======+
where XXXXXX_XXXXXXXX_XXXXXX_XXXXXXXX is a 32-bit unsigned integer
```

### Integer Family (`INTP`, `NINTP`, `UINT8`, `UINT16`, `UINT32`, `UINT64`, `INT8`, `INT16`, `INT32`, `INT64`)

`INT` format stores a signed or unsigned integer in 1, 2, 3, 5, or 9 bytes.

```
INTP stores 5-bit positive integer
+--------+
|101XXXXX|
+--------+
where XXXXX is a 5-bit unsigned integer

NINTP stores 5-bit negative integer
+--------+
|111YYYYY|
+--------+
where YYYYY is a 5-bit unsigned integer


INT8 stores 8-bit signed integer
+--------+--------+
|  0xD1  |XXXXXXXX|
+--------+--------+
where XXXXXXXX is an 8-bit signed integer

INT16 stores 16-bit signed integer
+--------+--------+--------+
|  0xD2  |XXXXXXXX|XXXXXXXX|
+--------+--------+--------+
where XXXXXXXX_XXXXXXXX is an 16-bit unsigned integer

INT32 stores 32-bit signed integer
+--------+--------+--------+--------+--------+
|  0xD3  |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|
+--------+--------+--------+--------+--------+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is an 32-bit signed integer

INT64 stores 64-bit signed integer
+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|  0xD4  |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|
+--------+--------+--------+--------+--------+--------+--------+--------+--------+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is an 64-bit signed integer


UINT8 stores 8-bit unsigned integer
+--------+--------+
|  0xD1  |XXXXXXXX|
+--------+--------+
where XXXXXXXX is an 8-bit unsigned integer

UINT16 stores 16-bit unsigned integer
+--------+--------+--------+
|  0xD2  |XXXXXXXX|XXXXXXXX|
+--------+--------+--------+
where XXXXXXXX_XXXXXXXX is an 16-bit unsigned integer

UINT32 stores 32-bit unsigned integer
+--------+--------+--------+--------+--------+
|  0xD3  |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|
+--------+--------+--------+--------+--------+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is an 32-bit unsigned integer

UINT64 stores 64-bit unsigned integer
+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|  0xD4  |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|
+--------+--------+--------+--------+--------+--------+--------+--------+--------+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is an 64-bit unsigned integer
```

### Boolean Family (`FALSE`, `TRUE`)

`BOOL` format stores a true or false in one byte.

```
FALSE:
+--------+
|  0xC0  |
+--------+

TRUE:
+--------+
|  0xC1  |
+--------+
```

### Float Family (`FLOAT32`, `FLOAT64`)

```
FLOAT32 stores IEEE 754 single precision floating point number.
+--------+--------+--------+--------+--------+
|  0xD9  |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|
+--------+--------+--------+--------+--------+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is a big-endian IEEE 754 single precision float point number.

FLOAT64 stores IEEE 754 double precision floating point number.
+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|  0xDA  |XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|XXXXXXXX|
+--------+--------+--------+--------+--------+--------+--------+--------+--------+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is a big-endian
IEEE 754 double precision floating point number
```

### Extension Family (`EXT8`, `EXT16`, `EXT32`)

```
EXT8 stores a mapping of up to 255 bytes
+--------+--------+--------+=======+
|  0xDB  |XXXXXXXX|YYYYYYYY| bytes |
+--------+--------+--------+=======+
where XXXXXXXX is a 8-bit unsigned integer for length
of bytes and YYYYYYYY is the 8-bit extension code

EXT16 stores a mapping of up to 65,535 bytes
+--------+----------------+--------+=======+
|  0xDC  |XXXXXXXXXXXXXXXX|YYYYYYYY| bytes |
+--------+----------------+--------+=======+
where XXXXXXXX_XXXXXXXX is a 16-bit unsigned integer for length
of bytes and YYYYYYYY is the 8-bit extension code

EXT32 stores a mapping of up to 4,294,967,295 bytes
+--------+--------------------------------+--------+=======+
|  0xDD  |XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX|YYYYYYYY| bytes |
+--------+--------------------------------+--------+=======+
where XXXXXXXX_XXXXXXXX_XXXXXXXX_XXXXXXXX is a 32-bit unsigned integer for length
of bytes and YYYYYYYY is the 8-bit extension code
```

### Null Family (`NULL`)

`NULL` format stores a null/nil/none value in 1 byte.

```
+--------+
|  0xDF  |
+--------+
```

## Future Improvements
  
- Handling and logic of recognizing `MARRAY[*P and *8]` being converted to `ARRAY[*8]`
  when space would be saved by this operation. Storage will be saved when there is more
  than one `*8` element and will be done if there are more than 3.

- Create a Cython implementation of the specification.

- Benchmarking against small, medium, and large JSON objects as well as individual object
  types against Messagepack.

## Implementations

- [Python](https://github.com/SethMichaelLarson/mashpack)

## License

Apache-2.0


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.0] (2018-01-22)
### Added

- Add support for extensions using `ExtType`.
- Add the Python implementation of `mashpack.Packer`.
- Add the Python implementation of `mashpack.Unpacker`.
- Add the `ARRAY8`, `ARRAY16`, `ARRAY32` data types for typed arrays to save
  space on header information.
- Add the `MARRAYP`, `MARRAY8`, `MARRAY16`, and `MARRAY32` data types for
  mixed type arrays which are equivalent to MessagePack arrays.
- Add the `EXT16` data type.
- Add all the base data types: `STRP`, `MAPP`, `INTP`, `NINTP`, `ARRAYP`,
  `FALSE`, `TRUE`, `MAP8`, `MAP16`, `MAP32`, `STR8`, `STR16`, `STR32`,
  `ARRAY8`, `ARRAY16`, `ARRAY32`, `BIN8`, `BIN16`, `BIN32`, `INT8`, `INT16`,
  `INT32`, `INT64`, `UINT8`, `UINT16`, `UINT32`, `UINT64`, `FLOAT32`,
  `FLOAT64`, `EXT8`, `EXT32`, and `NULL`.

### Removed
- Removed the `ARRAYP` data type in favor of `MARRAYP`.
- Removed the `MATRIX16` and `MATRIX32` types in favor of extensions.

[1.0.0]: https://github.com/SethMichaelLarson/mashpack/compare/e094d8eef3c29acfd0201141703a22de52af2ba0...1.0.0


tests/__init__.py
# Copyright 2018 Seth Michael Larson
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


tests/conftest.py
# Copyright 2018 Seth Michael Larson
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import pytest
from mashpack._fallback import (
    Unpacker as PythonUnpacker, Packer as PythonPacker
)


@pytest.fixture(scope='session')
def packer_type():
    if os.environ.get('TEST_WITH_CYTHON') == 'true':
        raise NotImplementedError()
    else:
        return PythonPacker


@pytest.fixture(scope='session')
def unpacker_type():
    if os.environ.get('TEST_WITH_CYTHON') == 'true':
        raise NotImplementedError()
    else:
        return PythonUnpacker


@pytest.fixture(scope='function')
def unpacker(unpacker_type):
    return unpacker_type()


@pytest.fixture(scope='function')
def packer(packer_type):
    return packer_type()


tests/test_packer.py
import struct
from mashpack import ExtType


def test_pack_ext8(packer):
    assert packer.pack(ExtType(127, b'\x00' * 0xFF)) == b'\xDB\xFF\x7F' + (b'\x00' * 0xFF)


def test_pack_ext16(packer):
    assert packer.pack(ExtType(127, b'\x00' * 0x100)) == b'\xDC' + struct.pack('>H', 0x100) + b'\x7F' + (b'\x00' * 0x100)


def test_pack_ext32(packer):
    assert packer.pack(ExtType(127, b'\x00' * 0x10000)) == b'\xDD' + struct.pack('>I', 0x10000) + b'\x7F' + (b'\x00' * 0x10000)


mashpack/__init__.py
# Copyright 2018 Seth Michael Larson
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from collections import namedtuple

__all__ = [
    'Packer', 'Unpacker', 'ExtType',
    'pack', 'packb', 'unpack', 'unpackb',
    'dump', 'dumps', 'load', 'loads'
]


class ExtType(namedtuple('ExtType', ['code', 'data'])):
    def __new__(cls, code, data):
        if not isinstance(code, int):
            raise TypeError('code must be int')
        if not isinstance(data, bytes):
            raise TypeError('data must be bytes')
        if not 0 <= code <= 0x7F:
            raise ValueError('code must be 0 to 127')
        return super(ExtType, cls).__new__(cls, code, data)


from ._fallback import Packer, Unpacker, unpack, unpackb


def pack(o, stream, **kwargs):
    packer = Packer(**kwargs)
    stream.write(packer.pack(o))


def packb(o, **kwargs) -> bytes:
    packer = Packer(**kwargs)
    return packer.pack(o)


# Compatibility with marshal/pickle
load = unpack
loads = unpackb
dump = pack
dumps = packb


mashpack/exceptions.py
# Copyright 2018 Seth Michael Larson
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


class BufferFull(Exception):
    pass


class OutOfData(Exception):
    pass


class ExtraData(ValueError):
    def __init__(self, unpacked, extra):
        self.unpacked = unpacked
        self.extra = extra

    def __str__(self):
        return "unpack(b) received extra data."


class PackValueError(ValueError):
    pass


class PackOverflowError(OverflowError, PackValueError):
    pass


mshpck.py
# Copyright 2018 Seth Michael Larson
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import struct
import typing

LARGEST_VARINT_COLLECTION = 1 * 1024 * 1024
ARRAY = 0x80
INT = 0xA0
UNUSED = 0xB
MAP = 0xC0
STR = 0xE0
FALSE = 0x00
TRUE = 0x10
FLOAT32 = 0x20
FLOAT64 = 0x30
NULL = 0x40
NEGINT = 0x50
BIN = 0x60
EXT = 0x70


def _packb_obj(obj, encoding='utf-8') -> bytes:
    if isinstance(obj, int):
        if obj > 0:
            return _encode_varint(obj, 3, INT)
        else:
            return _encode_varint(obj * -1, 4, NEGINT)
    elif isinstance(obj, str):
        return _encode_varint(len(obj), 3, STR) + obj.encode(encoding)
    elif obj is False:
        return b'\x00'
    elif obj is True:
        return b'\x10'
    elif isinstance(obj, float):
        return b'\x30' + struct.pack('!d', obj)
    elif obj is None:
        return b'\x40'
    elif isinstance(obj, bytes):
        return _encode_varint(len(obj), 4, BIN) + obj
    elif isinstance(obj, list):
        data = []
        for item in obj:
            data.append(_packb_obj(item, encoding))
        return _encode_varint(len(obj), 4, ARRAY) + b''.join(data)
    elif isinstance(obj, dict):
        data = []
        for key, value in obj.items():
            data.append(_packb_obj(key, encoding))
            data.append(_packb_obj(value, encoding))
        return _encode_varint(len(obj), 4, MAP) + b''.join(data)
    else:
        raise ValueError(f'Unknown object {obj!r}')


def packb(obj, encoding='utf-8') -> bytes:
    return _packb_obj(obj, encoding)


def unpackb(data: bytes, encoding='utf-8'):
    raise NotImplementedError()


def _encode_varint(varint: int, prefix: int, header: int=0) -> bytes:
    if varint < 0:
        raise TypeError('varint must be positive')

    # Single byte encoding
    if ((0x7F >> prefix) & varint) == varint:
        return int.to_bytes(header | varint | (0x80 >> prefix), 1, 'big')

    # Multiple byte encoding
    total = header | (varint & (0x7F >> prefix))
    offset = 1
    varint = varint >> (7 - prefix)
    while varint:
        total = (total << 8) | (varint & 0x7F)
        varint = varint >> 7
        offset += 1
    total |= 0x80

    # Packing data from integer into bytes
    data = []
    while offset >= 4:
        data.insert(0, int.to_bytes(total & 0xFFFFFFFF, 4, 'big'))
        total = total >> 32
        offset -= 4
    while offset > 0:
        data.insert(0, int.to_bytes(total & 0xFF, 1, 'big'))
        total = total >> 8
        offset -= 1

    return b''.join(data)


def _decode_varint(data: bytes, prefix: int, _collection=False) -> typing.Tuple[int, int]:
    total = 0
    offset = 0

    def _memory_check():
        if total and _collection and total > LARGEST_VARINT_COLLECTION:
                raise ValueError('Detected a large collection and '
                                 'failed decoding for safety purposes.')

    for i, val in enumerate(data):
        # Prefix byte
        if i == 0:
            total = (val & (0x7F >> prefix))
            offset = 7 - prefix
            _memory_check()
            if val & (1 << (7 - prefix)):
                break

        # Non-prefix byte
        else:
            total |= (val & 0x7F) << offset
            offset += 7
            _memory_check()
            if val & 0x80:
                break
    else:
        raise ValueError('No stop-bit was found for this varint')
    return total, i + 1


def testme(data):
    import msgpack
    data1 = packb(data)
    data2 = msgpack.packb(data)

    print('mshpck', data1, len(data1))
    print('msgpack', data2, len(data2))
